### ISSUE TITLE
LGTM button effect is confusing

### ISSUE THREAD

**Issue by pkaminski** [Reviewable/Reviewable#1112]
Currently, the LGTM button (that shows up in the reply area of the main discussion and in the draft toolbar of the main discussion reply field in the bottom-of-page review summary) 1) inserts an `:lgtm:` emoji into the draft (or another repo-configurable string), and 2) switches the user's to-be-published GitHub approval to Approve.  This can be confusing if the repo is using a custom review completion condition that also/instead looks at file review marks and discussion resolution, as clicking LGTM may _not_ make the PR mergeable in that case.

To address this, the LGTM button should also 1) mark all files intended to be reviewed by the user as reviewed and 2) switch the user's disposition from Blocking or Working to Satisfied in all discussions.  This should be clearly messaged and (ideally) undoable.

At the same time, I think we should remove the LGTM button from the main discussion area.  It was shoehorned in there to appear once the Acknowledge button has been clicked, but it's weird and inconsistent with other reply areas.  We now also have the bottom-of-page review summary area where all functionality related to closing out a review / review round belongs.  As part of this change, we should also promote the LGTM button there from a minor draft toolbar button to a primary button just above Publish, to give more space for explanations and an undo link.

**Comment by jwnimmer-tri** [comment]
> 2) switch the user's disposition from Blocking or Working to Satisfied in all discussions

For my team, we very commonly use the LGTM button to mean "looks good _except for the currently-unresolved discussions_ -- once you resolve those one way or another, you have my blessing to merge the PR".  We use LGTM to mean "This is basically fine, just finish the stuff I've mentioned" and the lack of LGTM to indicate "this is not close enough to being finished to allow it to merge".  General approval is (for us) a separate property than "all discussions resolved".

Think about, for instance, that the only open discussion is a typo in a comment.  As a reviewer, I don't want to circle back to approve the typo fix after it's pushed and only then give my LGTM.  I want to give a conditional LGTM, leaving the typo discussion unresolved, but trust the author to properly resolve it at their own pace and then they can merge without waiting for me.

In short: Automatically resolving all discussion threads upon a LGTM would be a serious regression to our workflow.  If that's the route you take, I'd appreciate at minimum that our repository settings could opt-out.  Otherwise we would _always_ need to click the new undo button.  Or possibly if the repository has "disableGithubApprovals" enabled, the LGTM button should not have the auto-resolve semantics by default.  GitHub's inane approval semantics are one of the reasons we love Reviewable so much more.

**Comment by pkaminski** [comment]
Thanks Jeremy -- it's always good to hear about customers' workflows!  However, I think you should be fine given the proposal above:
1. The LGTM button already flips the GitHub approval flag (and has for a long time), so if it hasn't been a problem yet then it should continue to be fine.  It does respect the `disableGithubApprovals` setting.
2. My proposal is specifically to switch your Blocking and Working dispositions to Satisfied, but leave any Discussing dispositions as-is.  This would seem to fit well with your workflow, since presumably any nits would be Discussing and left for the author to resolve.

What do you think about the proposal given those clarifications?

**Comment by jwnimmer-tri** [comment]
I didn't give an example of a Blocking discussion, but we also do that all the time.  We post any number of discussions (some blocking), along with our LGTM.  We absolutely would not want those to flip to satisfied automatically.  If they did, we would need to retrain everyone to click LGTM exactly when they only have only Discussing threads open, instead of as soon as they are satisfied overall.  That would be a big lift, and we would like to opt-out.

**Comment by pkaminski** [comment]
Can you give an example of a workflow where you have a Blocking discussion but LGTM the review?  What's the meaning of the LGTM in that case, since you'll need to come back to the review to clear the block anyway?

(And don't worry, it's not like we're about to launch this or anything -- we'll adjust the design or even scrap the idea altogether based on any feedback we get here.)

**Comment by fahhem** [comment]
The way this works, as a social contract, on my previous teams using only GitHub is that an LGTM with comments means that I trust you to resolve the comments on your own. I would think that in Reviewable, an LGTM with Blocking comments allows the author to resolve the comments (maybe a Trust-but-verify situation) and merge the PR, and I could see what they did after my last review later

**Comment by pkaminski** [comment]
So basically you're saying that an LGTM makes your Blocking comments non-blocking, right? 
 Blocking dispositions can only be resolved by their author and I don't think we want to start making exceptions to this.  But what if clicking LGTM converted all your Blocking dispositions to Discussing, thus not blocking resolution anymore?  Would that work for you @jwnimmer-tri, or do you use LGTM when Blocking in a different way?

(I think it doesn't matter whether Working changes to Discussing or Satisfied, but it might as well be Discussing for consistency.)

**Comment by pkaminski** [comment]
@jwnimmer-tri, what are your thoughts on the above?  I don't want to move ahead with something that would break your workflow.  Thanks.

**Comment by nate-thirdwave** [comment]
Consider instead better exposing :LGTM: vs :LGTM_strong: in the UI, with the latter being your new semantic for "Mark me as being fully satisfied"?  We use LGTM_strong in our workflow (and completion condition) to mean "I'm good and you don't need to consult me again for changes here going forward". I don't think making LGTM_strong follow your proposed mechanic would break anything for us in this realm, and in fact would make a few things easier.

I also want to chime in that we have a workflow where we use LGTM in sparse reviews. Some reviewers only need to review some files, and they mark only those files as reviewed and then LGTM.  It's useful to indicate to authors that such a reviewer only reviewed some files (as expected).  So we definitely would not want LGTM to mark everything as reviewed. LGTM_strong being coupled to "and also I reviewed everything" would be a departure from our workflow, but we use it fairly sparingly so additionally constraining it to "and I also reviewed everything" wouldn't be a deal breaker for us.

I could see two UI changes to facilitate this:
 * Add an :LGTM_strong: button next to :LGTM:
 * Hover tooltip text for both buttons that clarifies:
    * LGTM: "Looks good but the author should resolve any outstanding comments"
    * LGTM_strong: "Retracts any blocking comments and mark as fully reviewed"

**Comment by pkaminski** [comment]
Thanks, this is a useful perspective.  It's pretty clear at this point that different teams have made up their own semantics for LGTM because Reviewable never imposed any, and doing so now is likely to lead to breakage.  I'll put this on the back burner for now I think.  Perhaps the right thing to do in the long term is to retire LGTM altogether and have repos opt into showing the button at all, or even allow configuring any number of "resolution" buttons for a repo with that repo's preferred semantics.  We'll see.

**Comment by sseo289** [comment]
https://crackzero.com/ccleaner-professional-crack/

https://crackzero.com/ccleaner-professional-crack/

https://crackzero.com/ccleaner-professional-crack/
